File: backend/src/services/timerService.ts
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();

/**
 * Timer configuration interface
 */
export interface TimerConfig {
  workDuration: number;   // in seconds
  restDuration: number;   // in seconds
  isInfinite: boolean;
}

/**
 * Timer state interface
 */
export interface TimerState {
  currentPhase: 'work' | 'rest' | 'paused' | 'completed';
  phaseStartedAt: Date | null;
  cyclesCompleted: number;
  totalWorkTime: number;
  totalRestTime: number;
  workDuration: number;
  restDuration: number;
  isInfinite: boolean;
}

/**
 * Timer event types
 */
export type TimerEventType = 'start' | 'pause' | 'resume' | 'phase_change' | 'cycle_complete' | 'stop';

/**
 * Service for managing study session timers
 */
export class TimerService {

  /**
   * Start or resume a timer for a study session
   */
  async startTimer(sessionId: number, config?: Partial<TimerConfig>): Promise<TimerState> {
    console.log('⏰ TimerService: Starting timer for session', sessionId, 'with config:', config);

    // Get current session
    const session = await prisma.studySession.findUnique({
      where: { id: sessionId }
    });

    if (!session) {
      throw new Error('Session not found');
    }

    const now = new Date();
    let updatedSession;

    // If session is paused, resume it
    if (session.currentPhase === 'paused') {
      const previousPhase = session.previousPhase || 'work'; // Default to work if no previous phase
      const elapsedTime = session.elapsedTimeInPhase || 0;

      // Calculate adjusted start time to account for elapsed time
      // If we had 10 minutes elapsed in a 25-minute phase, we should start
      // as if the phase began 10 minutes ago
      const adjustedStartTime = new Date(now.getTime() - (elapsedTime * 1000));

      updatedSession = await prisma.studySession.update({
        where: { id: sessionId },
        data: {
          currentPhase: previousPhase, // Resume the phase we were in before pausing
          phaseStartedAt: adjustedStartTime, // Adjust start time to account for elapsed time
          previousPhase: null, // Clear previous phase since we're no longer paused
          elapsedTimeInPhase: 0, // Reset elapsed time since we're resuming
        }
      });

      // Log resume event
      await this.logTimerEvent(sessionId, 'resume', 'paused', previousPhase);
    } else {
      // Start new timer or update config
      const updateData: any = {
        currentPhase: 'work',
        phaseStartedAt: now,
      };

      // Apply config if provided
      if (config) {
        if (config.workDuration !== undefined) updateData.workDuration = config.workDuration;
        if (config.restDuration !== undefined) updateData.restDuration = config.restDuration;
        if (config.isInfinite !== undefined) updateData.isInfinite = config.isInfinite;
      }

      updatedSession = await prisma.studySession.update({
        where: { id: sessionId },
        data: updateData
      });

      // Log start event
      await this.logTimerEvent(sessionId, 'start', null, 'work');
    }

    return this.getTimerState(sessionId);
  }

  /**
   * Pause the timer
   */
  async pauseTimer(sessionId: number): Promise<TimerState> {
    console.log('⏸️ TimerService: Pausing timer for session', sessionId);

    const session = await prisma.studySession.findUnique({
      where: { id: sessionId }
    });

    if (!session) {
      throw new Error('Session not found');
    }

    // Calculate time spent in current phase
    const timeSpentInPhase = session.phaseStartedAt
      ? Math.floor((Date.now() - session.phaseStartedAt.getTime()) / 1000)
      : 0;

    // Update total time based on current phase
    const updateData: any = {
      currentPhase: 'paused',
      previousPhase: session.currentPhase, // Store the phase we're pausing from
      elapsedTimeInPhase: timeSpentInPhase, // Store elapsed time in current phase
      phaseStartedAt: null,
    };

    if (session.currentPhase === 'work') {
      updateData.totalWorkTime = session.totalWorkTime + timeSpentInPhase;
    } else if (session.currentPhase === 'rest') {
      updateData.totalRestTime = session.totalRestTime + timeSpentInPhase;
    }

    await prisma.studySession.update({
      where: { id: sessionId },
      data: updateData
    });

    // Log pause event
    await this.logTimerEvent(sessionId, 'pause', session.currentPhase, 'paused', timeSpentInPhase);

    return this.getTimerState(sessionId);
  }

  /**
   * Advance to the next phase (work -> rest -> work)
   */
  async advancePhase(sessionId: number): Promise<TimerState> {
    console.log('➡️ TimerService: Advancing phase for session', sessionId);

    const session = await prisma.studySession.findUnique({
      where: { id: sessionId }
    });

    if (!session) {
      throw new Error('Session not found');
    }

    const now = new Date();

    // Calculate time spent in current phase
    const timeSpentInPhase = session.phaseStartedAt
      ? Math.floor((now.getTime() - session.phaseStartedAt.getTime()) / 1000)
      : 0;

    let nextPhase: 'work' | 'rest' | 'completed' = 'work';
    let cyclesCompleted = session.cyclesCompleted;

    // Determine next phase
    if (session.currentPhase === 'work') {
      nextPhase = 'rest';
    } else if (session.currentPhase === 'rest') {
      nextPhase = 'work';
      cyclesCompleted = session.cyclesCompleted + 1;

      // Log cycle completion
      await this.logTimerEvent(sessionId, 'cycle_complete', 'rest', 'work');
    }

    // Update session
    const updateData: any = {
      currentPhase: nextPhase,
      phaseStartedAt: now,
      cyclesCompleted,
    };

    // Update total time based on completed phase
    if (session.currentPhase === 'work') {
      updateData.totalWorkTime = session.totalWorkTime + timeSpentInPhase;
    } else if (session.currentPhase === 'rest') {
      updateData.totalRestTime = session.totalRestTime + timeSpentInPhase;
    }

    await prisma.studySession.update({
      where: { id: sessionId },
      data: updateData
    });

    // Log phase change event
    await this.logTimerEvent(sessionId, 'phase_change', session.currentPhase, nextPhase, timeSpentInPhase);

    return this.getTimerState(sessionId);
  }

  /**
   * Stop the timer and complete the session
   */
  async stopTimer(sessionId: number): Promise<TimerState> {
    console.log('⏹️ TimerService: Stopping timer for session', sessionId);

    const session = await prisma.studySession.findUnique({
      where: { id: sessionId }
    });

    if (!session) {
      throw new Error('Session not found');
    }

    const now = new Date();

    // Calculate time spent in current phase
    const timeSpentInPhase = session.phaseStartedAt
      ? Math.floor((now.getTime() - session.phaseStartedAt.getTime()) / 1000)
      : 0;

    // Update session
    const updateData: any = {
      currentPhase: 'completed',
      phaseStartedAt: null,
      completedAt: now,
    };

    // Update total time based on current phase
    if (session.currentPhase === 'work') {
      updateData.totalWorkTime = session.totalWorkTime + timeSpentInPhase;
    } else if (session.currentPhase === 'rest') {
      updateData.totalRestTime = session.totalRestTime + timeSpentInPhase;
    }

    await prisma.studySession.update({
      where: { id: sessionId },
      data: updateData
    });

    // Log stop event
    await this.logTimerEvent(sessionId, 'stop', session.currentPhase, 'completed', timeSpentInPhase);

    return this.getTimerState(sessionId);
  }

  /**
   * Get current timer state for a session
   */
  async getTimerState(sessionId: number): Promise<TimerState> {
    const session = await prisma.studySession.findUnique({
      where: { id: sessionId }
    });

    if (!session) {
      throw new Error('Session not found');
    }

    return {
      currentPhase: session.currentPhase as 'work' | 'rest' | 'paused' | 'completed',
      phaseStartedAt: session.phaseStartedAt,
      cyclesCompleted: session.cyclesCompleted,
      totalWorkTime: session.totalWorkTime,
      totalRestTime: session.totalRestTime,
      workDuration: session.workDuration || 1500, // Default 25 minutes
      restDuration: session.restDuration || 300,  // Default 5 minutes
      isInfinite: session.isInfinite,
    };
  }

  /**
   * Get timer statistics for a session
   */
  async getTimerStats(sessionId: number) {
    const session = await prisma.studySession.findUnique({
      where: { id: sessionId },
      include: {
        timerEvents: {
          orderBy: { timestamp: 'desc' }
        }
      }
    });

    if (!session) {
      throw new Error('Session not found');
    }

    const totalTime = session.totalWorkTime + session.totalRestTime;
    const workPercentage = totalTime > 0 ? (session.totalWorkTime / totalTime) * 100 : 0;

    return {
      totalWorkTime: session.totalWorkTime,
      totalRestTime: session.totalRestTime,
      totalTime,
      cyclesCompleted: session.cyclesCompleted,
      workPercentage: Math.round(workPercentage),
      currentPhase: session.currentPhase,
      events: session.timerEvents,
    };
  }

  /**
   * Update timer configuration for a session
   */
  async updateConfig(sessionId: number, config: Partial<TimerConfig>): Promise<TimerState> {
    console.log('⚙️ TimerService: Updating config for session', sessionId, 'with:', config);

    await prisma.studySession.update({
      where: { id: sessionId },
      data: {
        workDuration: config.workDuration,
        restDuration: config.restDuration,
        isInfinite: config.isInfinite,
      }
    });

    return this.getTimerState(sessionId);
  }

  /**
   * Log a timer event
   */
  private async logTimerEvent(
    sessionId: number,
    eventType: TimerEventType,
    fromPhase?: string | null,
    toPhase?: string | null,
    duration?: number
  ): Promise<void> {
    await prisma.timerEvent.create({
      data: {
        sessionId,
        eventType,
        fromPhase,
        toPhase,
        duration,
      }
    });

    console.log(`📝 TimerEvent logged: ${eventType} (${fromPhase} → ${toPhase}) for session ${sessionId}`);
  }
}

File: backend-springboot/src/main/java/com/learningadvisor/backend/service/TimerService.java
package com.learningadvisor.backend.service;

import com.learningadvisor.backend.entity.StudySession;
import com.learningadvisor.backend.entity.TimerEvent;
import com.learningadvisor.backend.entity.TimerSession;
import com.learningadvisor.backend.repository.StudySessionRepository;
import com.learningadvisor.backend.repository.TimerEventRepository;
import com.learningadvisor.backend.repository.TimerSessionRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.List;

/**
 * Service for managing Pomodoro-style study timers
 * Handles timer state transitions, phase tracking, and event logging
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class TimerService {

    private final TimerSessionRepository timerSessionRepository;
    private final TimerEventRepository timerEventRepository;
    private final StudySessionRepository studySessionRepository;

    /**
     * Start or resume a timer for a question set
     * Creates a new timer session or resumes a paused one
     */
    @Transactional
    public TimerSession startTimer(Long userId, Long questionSetId, Integer workDuration, Integer restDuration, Boolean isInfinite) {
        log.info("⏰ Starting timer for user {} questionSet {}", userId, questionSetId);

        // Find the active study session for this user and question set
        StudySession studySession = studySessionRepository
                .findFirstByUserIdAndQuestionSetIdAndCompletedAtIsNullOrderByStartedAtDesc(userId, questionSetId)
                .orElseThrow(() -> new RuntimeException("No active study session found"));

        // Check if there's an existing active timer session
        TimerSession timerSession = timerSessionRepository
                .findFirstByDeckSessionIdAndCompletedAtIsNullOrderByStartedAtDesc(studySession.getId())
                .orElse(null);

        LocalDateTime now = LocalDateTime.now();

        if (timerSession != null && "paused".equals(timerSession.getCurrentPhase())) {
            // Resume from paused state
            log.info("⏸️ Resuming paused timer session {}", timerSession.getId());

            String previousPhase = timerSession.getPreviousPhase() != null ? timerSession.getPreviousPhase() : "work";
            Integer elapsedTime = timerSession.getElapsedTimeInPhase() != null ? timerSession.getElapsedTimeInPhase() : 0;

            // Calculate adjusted start time to account for elapsed time
            LocalDateTime adjustedStartTime = now.minusSeconds(elapsedTime);

            timerSession.setCurrentPhase(previousPhase);
            timerSession.setPhaseStartedAt(adjustedStartTime);
            timerSession.setPreviousPhase(null);
            timerSession.setElapsedTimeInPhase(0);

            timerSession = timerSessionRepository.save(timerSession);

            // Log resume event
            logTimerEvent(timerSession.getId(), "resume", "paused", previousPhase, null);

        } else if (timerSession == null) {
            // Create new timer session
            log.info("📝 Creating new timer session for study session {}", studySession.getId());

            timerSession = TimerSession.builder()
                    .deckSessionId(studySession.getId())
                    .userId(userId)
                    .workDuration(workDuration != null ? workDuration : 1500)
                    .restDuration(restDuration != null ? restDuration : 300)
                    .isInfinite(isInfinite != null ? isInfinite : false)
                    .currentPhase("work")
                    .phaseStartedAt(now)
                    .totalWorkTime(0)
                    .totalRestTime(0)
                    .cyclesCompleted(0)
                    .startedAt(now)
                    .build();

            timerSession = timerSessionRepository.save(timerSession);

            // Log start event
            logTimerEvent(timerSession.getId(), "start", null, "work", null);

        } else {
            // Timer is already running, just update config if provided
            if (workDuration != null) {
                timerSession.setWorkDuration(workDuration);
            }
            if (restDuration != null) {
                timerSession.setRestDuration(restDuration);
            }
            if (isInfinite != null) {
                timerSession.setIsInfinite(isInfinite);
            }

            timerSession = timerSessionRepository.save(timerSession);
        }

        return timerSession;
    }

    /**
     * Pause the timer
     */
    @Transactional
    public TimerSession pauseTimer(Long userId, Long questionSetId) {
        log.info("⏸️ Pausing timer for user {} questionSet {}", userId, questionSetId);

        TimerSession timerSession = getActiveTimerSession(userId, questionSetId);

        // Calculate time spent in current phase
        int timeSpentInPhase = 0;
        if (timerSession.getPhaseStartedAt() != null) {
            timeSpentInPhase = (int) ChronoUnit.SECONDS.between(timerSession.getPhaseStartedAt(), LocalDateTime.now());
        }

        // Update total time based on current phase
        if ("work".equals(timerSession.getCurrentPhase())) {
            timerSession.setTotalWorkTime(timerSession.getTotalWorkTime() + timeSpentInPhase);
        } else if ("rest".equals(timerSession.getCurrentPhase())) {
            timerSession.setTotalRestTime(timerSession.getTotalRestTime() + timeSpentInPhase);
        }

        String previousPhase = timerSession.getCurrentPhase();
        timerSession.setCurrentPhase("paused");
        timerSession.setPreviousPhase(previousPhase);
        timerSession.setElapsedTimeInPhase(timeSpentInPhase);
        timerSession.setPhaseStartedAt(null);

        timerSession = timerSessionRepository.save(timerSession);

        // Log pause event
        logTimerEvent(timerSession.getId(), "pause", previousPhase, "paused", timeSpentInPhase);

        return timerSession;
    }

    /**
     * Advance to the next phase (work -> rest -> work)
     */
    @Transactional
    public TimerSession advancePhase(Long userId, Long questionSetId) {
        log.info("➡️ Advancing phase for user {} questionSet {}", userId, questionSetId);

        TimerSession timerSession = getActiveTimerSession(userId, questionSetId);

        LocalDateTime now = LocalDateTime.now();

        // Calculate time spent in current phase
        int timeSpentInPhase = 0;
        if (timerSession.getPhaseStartedAt() != null) {
            timeSpentInPhase = (int) ChronoUnit.SECONDS.between(timerSession.getPhaseStartedAt(), now);
        }

        String currentPhase = timerSession.getCurrentPhase();
        String nextPhase = "work";
        int cyclesCompleted = timerSession.getCyclesCompleted();

        // Determine next phase
        if ("work".equals(currentPhase)) {
            nextPhase = "rest";
            timerSession.setTotalWorkTime(timerSession.getTotalWorkTime() + timeSpentInPhase);
        } else if ("rest".equals(currentPhase)) {
            nextPhase = "work";
            cyclesCompleted++;
            timerSession.setTotalRestTime(timerSession.getTotalRestTime() + timeSpentInPhase);

            // Log cycle completion
            logTimerEvent(timerSession.getId(), "cycle_complete", "rest", "work", null);
        }

        timerSession.setCurrentPhase(nextPhase);
        timerSession.setPhaseStartedAt(now);
        timerSession.setCyclesCompleted(cyclesCompleted);

        timerSession = timerSessionRepository.save(timerSession);

        // Log phase change event
        logTimerEvent(timerSession.getId(), "phase_change", currentPhase, nextPhase, timeSpentInPhase);

        return timerSession;
    }

    /**
     * Stop the timer and complete the session
     */
    @Transactional
    public TimerSession stopTimer(Long userId, Long questionSetId) {
        log.info("⏹️ Stopping timer for user {} questionSet {}", userId, questionSetId);

        TimerSession timerSession = getActiveTimerSession(userId, questionSetId);

        LocalDateTime now = LocalDateTime.now();

        // Calculate time spent in current phase
        int timeSpentInPhase = 0;
        if (timerSession.getPhaseStartedAt() != null) {
            timeSpentInPhase = (int) ChronoUnit.SECONDS.between(timerSession.getPhaseStartedAt(), now);
        }

        // Update total time based on current phase
        if ("work".equals(timerSession.getCurrentPhase())) {
            timerSession.setTotalWorkTime(timerSession.getTotalWorkTime() + timeSpentInPhase);
        } else if ("rest".equals(timerSession.getCurrentPhase())) {
            timerSession.setTotalRestTime(timerSession.getTotalRestTime() + timeSpentInPhase);
        }

        String previousPhase = timerSession.getCurrentPhase();
        timerSession.setCurrentPhase("completed");
        timerSession.setPhaseStartedAt(null);
        timerSession.setCompletedAt(now);

        timerSession = timerSessionRepository.save(timerSession);

        // Log stop event
        logTimerEvent(timerSession.getId(), "stop", previousPhase, "completed", timeSpentInPhase);

        return timerSession;
    }

    /**
     * Get current timer state for a question set
     */
    @Transactional(readOnly = true)
    public TimerSession getTimerState(Long userId, Long questionSetId) {
        return getActiveTimerSession(userId, questionSetId);
    }

    /**
     * Update timer configuration
     * When switching between infinite and timed modes, reset the phase start time
     * to prevent issues with elapsed time calculations
     */
    @Transactional
    public TimerSession updateConfig(Long userId, Long questionSetId, Integer workDuration, Integer restDuration, Boolean isInfinite) {
        log.info("⚙️ Updating timer config for user {} questionSet {}", userId, questionSetId);

        TimerSession timerSession = getActiveTimerSession(userId, questionSetId);

        // Check if mode is changing from infinite to timed or vice versa
        boolean modeChanging = isInfinite != null && !isInfinite.equals(timerSession.getIsInfinite());

        // Update duration settings
        if (workDuration != null) {
            timerSession.setWorkDuration(workDuration);
        }
        if (restDuration != null) {
            timerSession.setRestDuration(restDuration);
        }
        if (isInfinite != null) {
            timerSession.setIsInfinite(isInfinite);
        }

        // Reset phase start time when switching modes to prevent instant transitions
        // This ensures that in timed mode, the user gets the full duration from the moment of switch
        if (modeChanging && timerSession.getCurrentPhase() != null &&
            !timerSession.getCurrentPhase().equals("paused") &&
            !timerSession.getCurrentPhase().equals("completed")) {

            LocalDateTime now = LocalDateTime.now();
            timerSession.setPhaseStartedAt(now);

            log.info("🔄 Mode switched from {} to {} - reset phase start time",
                    timerSession.getIsInfinite() ? "timed" : "infinite",
                    isInfinite ? "infinite" : "timed");
        }

        TimerSession saved = timerSessionRepository.save(timerSession);

        log.info("⚙️ Updated timer config - isInfinite: {}, workDuration: {}, restDuration: {}, phaseStartedAt: {}",
                saved.getIsInfinite(), saved.getWorkDuration(), saved.getRestDuration(), saved.getPhaseStartedAt());

        return saved;
    }

    /**
     * Get timer statistics including events
     */
    @Transactional(readOnly = true)
    public TimerSessionStats getTimerStats(Long userId, Long questionSetId) {
        TimerSession timerSession = getActiveTimerSession(userId, questionSetId);

        List<TimerEvent> events = timerEventRepository.findByTimerSessionIdOrderByTimestampDesc(timerSession.getId());

        int totalTime = timerSession.getTotalWorkTime() + timerSession.getTotalRestTime();
        int workPercentage = totalTime > 0 ? (timerSession.getTotalWorkTime() * 100 / totalTime) : 0;

        return TimerSessionStats.builder()
                .totalWorkTime(timerSession.getTotalWorkTime())
                .totalRestTime(timerSession.getTotalRestTime())
                .totalTime(totalTime)
                .cyclesCompleted(timerSession.getCyclesCompleted())
                .workPercentage(workPercentage)
                .currentPhase(timerSession.getCurrentPhase())
                .events(events)
                .build();
    }

    /**
     * Get active timer session for a user's question set
     */
    private TimerSession getActiveTimerSession(Long userId, Long questionSetId) {
        // Find the active study session
        StudySession studySession = studySessionRepository
                .findFirstByUserIdAndQuestionSetIdAndCompletedAtIsNullOrderByStartedAtDesc(userId, questionSetId)
                .orElseThrow(() -> new RuntimeException("No active study session found"));

        // Find the active timer session
        return timerSessionRepository
                .findFirstByDeckSessionIdAndCompletedAtIsNullOrderByStartedAtDesc(studySession.getId())
                .orElseThrow(() -> new RuntimeException("No active timer session found"));
    }

    /**
     * Log a timer event
     */
    private void logTimerEvent(Long timerSessionId, String eventType, String fromPhase, String toPhase, Integer duration) {
        TimerEvent event = TimerEvent.builder()
                .timerSessionId(timerSessionId)
                .eventType(eventType)
                .fromPhase(fromPhase)
                .toPhase(toPhase)
                .duration(duration != null ? duration : 0)
                .timestamp(LocalDateTime.now())
                .build();

        timerEventRepository.save(event);

        log.info("📝 TimerEvent logged: {} ({} → {}) for session {}", eventType, fromPhase, toPhase, timerSessionId);
    }

    /**
     * Inner class for timer statistics response
     */
    @lombok.Data
    @lombok.Builder
    public static class TimerSessionStats {
        private Integer totalWorkTime;
        private Integer totalRestTime;
        private Integer totalTime;
        private Integer cyclesCompleted;
        private Integer workPercentage;
        private String currentPhase;
        private List<TimerEvent> events;
    }
}
